tailscale/tailscale
# cmd/k8s-operator,k8s-operator/sessionrecording: support recording kubectl exec sessions over WebSockets

Kubernetes currently supports two streaming protocols, SPDY and WebSockets.
WebSockets are replacing SPDY, see https://github.com/kubernetes/enhancements/issues/4006.

We were currently only supporting SPDY, erroring out if session was not SPDY and relying on the kube's built-in [SPDY fallback](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets#proposal-kubectl-websocket-executor-and-fallback-executor) behaviour.
    
This PR:
    
 - adds support for parsing contents of 'kubectl exec' sessions streamed
    over WebSockets

 - adds logic to distinguish 'kubectl exec' requests for a SPDY/WebSockets
    sessions and call the relevant handler

## Notes

- The message parsing functionality in this PR parses the raw bytes as WebSockets messages according to [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455). I had a look at using [gorilla/websockets](), [nyoor.io/websocket](https://pkg.go.dev/nhooyr.io/websocket) or [x/net/websocket](https://pkg.go.dev/golang.org/x/net/websocket), but all three of those assume running as one end that fully owns the TCP connection and treats the WebSocket messages in opinionated way, so wasn't suited for intercepting middleware like our proxy.

- To test this out:

  1. ensure that you are using kubectl and kube API server 1.30 or newer  
  2. deploy the operator with API server proxy enabled and set up session recording
  3. you can then run `kubectl exec -v=5 ...` to get more verbose output to verify that the connection is established over WebSockets i.e
 ```
$ kubectl -v 5 exec -ti proxy -- sh
I0729 12:46:59.175384  431894 podcmd.go:88] Defaulting container name to nettools
I0729 12:46:59.345479  431894 websocket.go:133] The subprotocol is v5.channel.k8s.io
                                                                                    I0729 12:46:59.345752  431894 websocket.go:357] Write() on stream 4
                                                                                                                                                       I0729 12:46:59.345846  431894 websocket.go:390] Write() done on stream 4
     / # I0729 12:46:59.407274  431894 websocket.go:357] Write() on stream 0
                                                                            I0729 12:46:59.407373  431894 websocket.go:390] Write() done on stream 0...
```
- From not very thorough performance testing it seems like the websocket exec with recording is more latent than the same, but over SPDY. I haven't yet done thorough testing to verify where the latency comes from, this is a TODO

- Currently tsrecorder's built-in record player will not often not be able to play the recording because of a race between receiving the initial resize message with terminal size from client and receiving the first message to be recorded from server. This is fixed in https://github.com/tailscale/tailscale/pull/12965
    
Updates tailscale/corp#19821
